### ISSUE TITLE
koreader crashes on large manga chapters

### ISSUE THREAD

**Issue by KeePach** [hanatsumi/rakuyomi#24]
when selecting a source with a large number of chapters like one piece, the plugin crashes.

**To Reproduce**
when selecting one piece from ComicK (or TCB scans), the plugin says "refreshing chapters" and then nothing happens, exiting the plugin and entering again causes the crash to happen

**Expected behavior**
To see the chapters

**Environment:**

 - Device: [e.g. Kindle Paperwhite 3]
 - KOReader version: [e.g. 2024.07]
 - Source name: [ComicK, TCB scans]

it is my assumption that this is caused due the large amount of chapters becuase when opening another manga that's not one piece, this error didn't happened.

[crash.log](https://github.com/user-attachments/files/16821591/crash.log)

Also, while I'm here, I would like to ask for a feature that when you have mangas on your library, would it be possible to see where they're being sourced? like NameofManga (ComicK), NameofManga (komga), etc. Oh, and also a way to deleted them from my saved mangas, (or at least I can't seem to find how to delete saved mangas from my library).

May I reccomend opening a thread on mobileread forums, it would be beneficial for the plugin for more people to know and test it.

**Comment by hanatsumi** [comment]
Hey! Looking from the logs, I'd assume the program ran out of memory while reading the chapter list:
```
08/30/24-13:52:29 INFO  Requesting to  {
  authority = "localhost:30727",
  host = "localhost",
  path = "/mangas/en.tcbscans/%2Fmangas%2F5%2Fone-piece/refresh-chapters",
  port = "30727",
  scheme = "http"
} --[[table: 0x40777fe8]]  
memory allocation of 1268608 bytes failed
```

I'll attempt to replicate this in a real Kindle device, as I couldn't really reproduce it while running it in on a desktop (which, well, has way more memory available).

> Also, while I'm here, I would like to ask for a feature that when you have mangas on your library, would it be possible to see where they're being sourced? like NameofManga (ComicK), NameofManga (komga), etc. Oh, and also a way to deleted them from my saved mangas, (or at least I can't seem to find how to delete saved mangas from my library)

Sure!

> May I reccomend opening a thread on mobileread forums, it would be beneficial for the plugin for more people to know and test it.

Alright, I'll do it later

**Comment by hanatsumi** [comment]
Hi, a quick update about this issue: I've identified that the memory usage issues stem mostly from attempting to parse the chapter data read from the sources. Each of the reported sources parses data in a different way: ComicK reads its data from a JSON API, which TCB Scans attempts to parse a HTML page.

I've pushed some memory usage fixes related to JSON parsing in #26, which should make reading from JSON sources a lot less memory-hungry. In my tests with a real device (a Kindle Paperwhite 5), now it's possible to load the chapter listing for One Piece from ComicK.

I'll work on some more fixes for HTML-parsing sources, but now at least one of the sources is usable :sweat_smile:
